### ISSUE TITLE
vispy.gloo.wrappers.read_pixels is too slow

### ISSUE THREAD

**Issue by ZuchaoWang** [vispy/vispy#733]
Hi, currently I'm trying to do some offline reading. I bascally draw something in a fbo, then read the pixels out, save that to a numpy array:

    rendertex = gloo.Texture2D(format='rgba',shape=(600,800,4))
    fbo = gloo.FrameBuffer(rendertex)

    im = None
    with fbo:
        # draw something
        im = fbo.read()

    fbo.delete()
    rendertex.delete()

But I find the function _fbo.read()_ to be very slow. The slowness seems due to this function call: _vispy.gloo.wrappers.read\_pixels_. Someone says that read_pixels will block all other GL functions, and some suggest to use PBO as a fix (with C API): http://www.mathematik.uni-dortmund.de/~goeddeke/gpgpu/tutorial3.html

I wonder if there's any possibility to make it faster, by either modifying the vispy library, or offering some other quick fixes. Thanks.

**Comment by rossant** [comment]
How slow is it for you? How does it compare to average CPU-GPU transfer rates?

**Comment by ZuchaoWang** [comment]
I'm using a win7 OS, with NVIDIA GeForce GT 330M GPU. With ipython notebook, for 2 million random points, the following code takes 300 ms:

    with fbo:
        # draw the points
        im = fbo.read()

but the next code only takes 10 ms:

    with fbo:
        # draw the points

**Comment by almarklein** [comment]
The reading of pixels in inherently notoriously slow. Nothing we can do about that :/

**Comment by ZuchaoWang** [comment]
Yes, read_pixels is an OpenGL function. But I find there are some alternative ways to read pixels much faster. I wonder whether it is possible to implement them in the vispy project, and how difficult it is?

https://vec.io/posts/faster-alternatives-to-glreadpixels-and-glteximage2d-in-opengl-es
http://www.mathematik.uni-dortmund.de/~goeddeke/gpgpu/tutorial3.html

**Comment by djhoese** [comment]
This is a very interesting idea (hello 5 years later). I'm not sure how much performance gain you would get since (as the goeddeke page mentions) this is still synchronous on the GPU side and is still dependent on the transer speed of your GPU to CPU memory. The benefit seems to be in the cases where you can say "OpenGL map this GPU buffer to memory in the CPU, now I'll do other stuff, now I want the data". In python I think these three parts would all have to happen at the same time to provide a simple enough interface. Unless the interface involves "request GPU data" and "now I want GPU data as a numpy array", but that is still very low-level.

@rougier Any experience with PBOs or EGLImage extension talked about in these links?

**Comment by rougier** [comment]
Sorry, no experience with EGLImage.
